src/middlewared/middlewared/plugins/nvmet/kernel.py

```diff
@@ -0,0 +1,790 @@
+                    attrs = cls.map_port_to_referral_attrs(port, render_ctx, parent_index == index)
+                    attrs['enable'] = '1'
+                    path = pathlib.Path(referrals_path, key)
+                    retries = cls.set_mapped_attrs(path, attrs, retries, render_ctx)
+
+        yield
+
+        for path in to_remove:
+            path.rmdir()
+
+
+class NvmetPortAnaReferralConfig(NvmetPortReferralConfig):
+    directory = 'ports'
+    query = 'nvmet.port.usage'
+    query_key = 'ana_referrals'
+
+    @classmethod
+    def handle_port(cls, index):
+        return index >= ANA_PORT_INDEX_OFFSET
+
+    @classmethod
+    def index(cls, index):
+        if index < ANA_PORT_INDEX_OFFSET:
+            return index + ANA_PORT_INDEX_OFFSET
+        else:
+            return index
+
+    @classmethod
+    def ports_by_index(cls, render_ctx):
+        # We want to update the index to the ANA specific one.
+        # This will then be used by the map_port_to_referral_attrs in
+        # this class to work out the address to be used.
+        result = {}
+        for port in render_ctx['nvmet.port.query']:
+            new_index = cls.index(port['index'])
+            if new_index != port['index']:
+                result[new_index] = port.copy() | {'index': new_index}
+            else:
+                result[new_index] = port
+        return result
+
+    @classmethod
+    def map_port_to_referral_attrs(cls, attrs: dict, render_ctx: dict, remote: bool):
+        data = super().map_port_to_referral_attrs(attrs, render_ctx, remote)
+        # This is an ANA port, update the addr_traddr
+        # We could be pointing at other ports on the same node, or on the
+        # other node - specified by remote parameter
+        if attrs.get('index', 0) > ANA_PORT_INDEX_OFFSET:
+            curval = attrs['addr_traddr']
+            prefix = attrs['addr_trtype'].lower()
+            choices = render_ctx[f'{prefix}.nvmet.port.transport_address_choices']
+            pair = choices[curval].split('/')
+            match render_ctx['failover.node']:
+                case 'A':
+                    if remote:
+                        newval = pair[1]
+                    else:
+                        newval = pair[0]
+                case 'B':
+                    if remote:
+                        newval = pair[0]
+                    else:
+                        newval = pair[1]
+            data['addr_traddr'] = newval
+        return data
+
+
+class NvmetSubsysConfig(NvmetConfig):
+    directory = 'subsystems'
+    query = 'nvmet.subsys.query'
+    query_key = 'subnqn'
+
+    @classmethod
+    def pre_delete(cls, path: pathlib.Path, render_ctx: dict):
+        # If we are force deleting a subsystem, then namespaces
+        # will have been deleted from the config, but not yet
+        # propagated live.  So, delete the namespaces here.
+        for ns in pathlib.Path(path / 'namespaces').iterdir():
```

Comment:
Again, `shutil.rmtree` should be investigated to remove so many atomic `rmdir` operations.